simpleflow/process/supervisor.py

```diff
@@ -68,5 +80,48 @@ def target(self):
             )
             child.start()
             self._processes.append(child)
+
+        # wait for all processes to finish
         for proc in self._processes:
             proc.join()
+
+    def bind_signal_handlers(self):
+        """
+        Binds signals for graceful shutdown:
+        - SIGTERM and SIGINT lead to a graceful shutdown
+        - other signals are ignored for now
+        """
+        # NB: Function is nested to have a reference to *self*.
+        def _handle_graceful_shutdown(signum, frame):
+            signals_map = { 2: "SIGINT", 15: "SIGTERM" }
+            signal_name = signals_map.get(signum, signum)
+            logger.info("process: caught signal signal={} pid={}".format(
+                signal_name, os.getpid()))
+            self.terminate()
+
+        # bind SIGTERM and SIGINT
+        signal.signal(signal.SIGTERM, _handle_graceful_shutdown)
+        signal.signal(signal.SIGINT, _handle_graceful_shutdown)
+
+    def terminate(self):
+        """
+        Terminate all worker processes managed by this Supervisor.
+        """
+        logger.error("processes: {}".format(self._processes))
+        if self._terminating:
+            logger.info("process: already shutting down...")
+            return
+        self._terminating = True
+        logger.info(
+            "process: will stop workers, this might take up several minutes. "
+            "Please, be patient."
+        )
+        self._killall()
+
+    def _killall(self):
+        """
+        Sends a stop (SIGTERM) signal to all worker processes.
+        """
+        for child in self._processes:
+            logger.info("process: sending SIGTERM to pid={}".format(child.pid))
+            os.kill(child.pid, signal.SIGTERM)
```

Comment:
After which timeout? Here we work with the assumption that activity tasks may be arbitrarily long ; a SIGTERM is supposed to ask the worker not to take new task anymore and wait for the current task to finish, but in some cases it can take 4, 5, 6 hours.